_posts/2025-06-24-kafka-queue.md

```diff
@@ -0,0 +1,704 @@
+
+        KafkaShareConsumer<String, String> consumer = new KafkaShareConsumer<>(consumerProps);
+
+        consumer.subscribe(List.of("kafka-queue-topic"));
+
+        while (true) {
+            ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(1000));
+
+            for (ConsumerRecord<String, String> record : records) {
+                if (record.key() != null && "skip".equals(record.key())) {  // #2 
+                    System.out.println("Skipping the record - Offset: " + record.offset() +
+                            " Key: " + record.key() +
+                            " Value: " + record.value() +
+                            " Delivery count: " + record.deliveryCount().get());
+                    consumer.acknowledge(record, AcknowledgeType.RELEASE);
+                    break; // #3
+                } else {
+                    try {
+                        processRecord(record);
+                    } catch (JsonProcessingException e) {
+                        System.out.println("Failed to process the record- Offset: " + record.offset() + " Value: " + record.value());
+                        handlePoisonRecord(record, producer); // #4
+                        consumer.acknowledge(record, AcknowledgeType.REJECT);  // #5
+                    } catch (Exception e ) {
+                        System.out.println("Failed to process the record- Offset: " + record.offset() + " Value: " + record.value());
+                        consumer.acknowledge(record, AcknowledgeType.RELEASE); // #6
+                    }
+                }
+                
+                consumer.acknowledge(record, AcknowledgeType.ACCEPT); // #7
+            }
+            consumer.commitSync(); // #7
+        }
+    }
+
+    private static void processRecord(ConsumerRecord<String, String> record) throws JsonProcessingException {
+        JsonNode json = OBJECT_MAPPER.readTree(record.value());
+        System.out.println("Processed record with offset " + record.offset() + ": " + json.toPrettyString());
+    }
+
+    private static void handlePoisonRecord(ConsumerRecord<String, String> record, KafkaProducer<String, String> producer) {
+        producer.send(new ProducerRecord<>("dead-letter-queue-topic", record.key(), record.value()));
+    }
+}
+
+```
+
+#1. In this example, I have used the new consumer configuration added for this feature, `share.acknowledgement.mode` and set it to `explicit`. This configuration is set to `implicit` by default, which is why I didn’t need to set this configuration for the previous example. 
+
+#2. The consumer immediately releases a record if its key is set to `skip`, just to demonstrate the retry of delivery attempts. The record is fetched again in the next call to `poll()` and eventually hits the maximum delivery attempts. Then it is transitioned into <b>Archived</b> state because the client never processes it. Once archived, this record is no longer available for another delivery. 
```

Comment:
```suggestion
#2. The consumer immediately releases a record if its key is set to `skip`, just to demonstrate the retry of delivery attempts. The record is fetched again in the next call to `poll()` and eventually hits the maximum delivery attempts. Then it is transitioned into an <b>Archived</b> state because the client never processes it. Once archived, this record is no longer available for another delivery. 
```